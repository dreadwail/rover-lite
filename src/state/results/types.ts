export interface Badge {
  readonly content: string;
  readonly imageUrl: string;
  readonly title: string;
}

export interface Listing {
  readonly thumbnailUrl: string;
  readonly rank: number;
  readonly name: string;
  readonly badges: Badge[];
  readonly tagline: string;
  readonly neighborhood: string;
  readonly city: string;
  readonly state: string;
  readonly zip: string;
  readonly repeatClientCount: number;
  readonly reviewCount: number;
  readonly ratingAverage: number;
  readonly featuredReview: string;
  readonly currencySymbol: string;
  readonly price: number;
  readonly priceUnit: string;
  readonly webUrl: string;
}

export interface Results {
  readonly isFetching: boolean;
  readonly hasFailed: boolean;
  readonly isStale: boolean;
  readonly listings: Listing[];
}
