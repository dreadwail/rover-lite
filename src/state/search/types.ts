import { SearchType } from '../searchTypes/types';

export type SearchInputs = Record<string, any>;

export interface Search {
  readonly searchType: SearchType;
  readonly latitude: number;
  readonly longitude: number;
  readonly inputs: SearchInputs;
}
