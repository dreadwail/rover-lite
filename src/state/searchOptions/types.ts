export interface OptionCategory {
  readonly categoryId: string;
  readonly rank: number;
  readonly displayName: string;
}

export interface BasicOption {
  readonly type: string;
  readonly required: boolean;
  readonly label: string;
  readonly category: OptionCategory;
  readonly rank: number;
  readonly helpText?: string;
}

export interface DateOption extends BasicOption {
  readonly type: 'date';
}

export interface Choice {
  readonly value: any;
  readonly displayName: string;
}

export interface ChoiceOption extends BasicOption {
  readonly type: 'choice';
  readonly choices: Choice[];
}

export interface MultipleChoiceOption extends BasicOption {
  readonly type: 'multiple choice';
  readonly choices: Choice[];
}

export interface IntegerOption extends BasicOption {
  readonly type: 'integer';
  readonly minValue: number;
  readonly maxValue: number;
}

export interface BooleanOption extends BasicOption {
  readonly type: 'boolean';
}

export type Option =
  | BasicOption
  | DateOption
  | ChoiceOption
  | MultipleChoiceOption
  | IntegerOption
  | BooleanOption;

export const isBooleanOption = (option: Option): option is BooleanOption =>
  option.type === 'boolean';
export const isIntegerOption = (option: Option): option is IntegerOption =>
  option.type === 'integer';
export const isMultipleChoiceOption = (option: Option): option is MultipleChoiceOption =>
  option.type === 'multiple choice';
export const isChoiceOption = (option: Option): option is ChoiceOption => option.type === 'choice';
export const isDateOption = (option: Option): option is DateOption => option.type === 'date';
export const isBasicOption = (option: Option): option is BasicOption =>
  [isBooleanOption, isIntegerOption, isChoiceOption, isMultipleChoiceOption, isDateOption].every(
    check => !check(option)
  );

export type Options = Record<string, Option>;
export type OptionsByCategoryId = Record<string, Options>;

export interface SearchOptions {
  readonly isFetching: boolean;
  readonly hasFailed: boolean;
  readonly options: Options;
}
