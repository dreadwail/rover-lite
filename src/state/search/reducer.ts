import { Reducer } from 'redux';

import { SearchFieldChangedAction, SearchTypeChangedAction } from '../../domains/search/actions';

import { Search } from './types';

export type SearchAction = SearchTypeChangedAction | SearchFieldChangedAction;

const hardCodedLatitude = 47.6843725;
const hardCodedLongitude = -122.3874837;

const initialState: Search = {
  searchType: 'overnight-boarding',
  latitude: hardCodedLatitude,
  longitude: hardCodedLongitude,
  inputs: {},
};

const reducer: Reducer<Search, SearchAction> = (state = initialState, action) => {
  switch (action.type) {
    case 'SEARCH.TYPE_CHANGED':
      return { ...initialState, searchType: action.payload };
    case 'SEARCH.FIELD_CHANGED':
      return {
        ...state,
        inputs: {
          ...state.inputs,
          [action.payload.key]: action.payload.value,
        },
      };
    default: {
      return state;
    }
  }
};

export default reducer;
