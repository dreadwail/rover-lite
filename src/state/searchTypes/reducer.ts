import { Reducer } from 'redux';

import {
  FetchSearchTypesFulfilledAction,
  FetchSearchTypesPendingAction,
  FetchSearchTypesRejectedAction,
} from '../../domains/searchTypes/actions';

import { SearchTypes } from './types';

export type FetchSearchTypesAction =
  | FetchSearchTypesPendingAction
  | FetchSearchTypesFulfilledAction
  | FetchSearchTypesRejectedAction;

const initialState: SearchTypes = {
  isFetching: false,
  hasFailed: false,
  endpoints: {},
};

const reducer: Reducer<SearchTypes, FetchSearchTypesAction> = (state = initialState, action) => {
  switch (action.type) {
    case 'FETCH_SEARCH_TYPES.PENDING':
      return { ...state, isFetching: true, hasFailed: false };
    case 'FETCH_SEARCH_TYPES.FULFILLED':
      return { ...state, isFetching: false, hasFailed: false, endpoints: action.payload };
    case 'FETCH_SEARCH_TYPES.REJECTED':
      return { ...state, isFetching: false, hasFailed: true };
    default: {
      return state;
    }
  }
};

export default reducer;
