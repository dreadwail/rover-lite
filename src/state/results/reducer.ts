import { Reducer } from 'redux';

import {
  FetchResultsFulfilledAction,
  FetchResultsPendingAction,
  FetchResultsRejectedAction,
} from '../../domains/results/actions';
import { SearchFieldChangedAction, SearchTypeChangedAction } from '../../domains/search/actions';

import { Results } from './types';

export type ResultsAction =
  | FetchResultsPendingAction
  | FetchResultsFulfilledAction
  | FetchResultsRejectedAction
  | SearchTypeChangedAction
  | SearchFieldChangedAction;

const initialState: Results = {
  isFetching: false,
  hasFailed: false,
  isStale: true,
  listings: [],
};

const reducer: Reducer<Results, ResultsAction> = (state = initialState, action) => {
  switch (action.type) {
    case 'FETCH_RESULTS.PENDING':
      return { ...state, isFetching: true, hasFailed: false, isStale: false };
    case 'FETCH_RESULTS.FULFILLED':
      return {
        ...state,
        isFetching: false,
        hasFailed: false,
        isStale: false,
        listings: action.payload,
      };
    case 'FETCH_RESULTS.REJECTED':
      return { ...state, isFetching: false, hasFailed: true };
    case 'SEARCH.TYPE_CHANGED':
      return { ...state, isStale: true };
    case 'SEARCH.FIELD_CHANGED':
      return { ...state, isStale: true };
    default: {
      return state;
    }
  }
};

export default reducer;
