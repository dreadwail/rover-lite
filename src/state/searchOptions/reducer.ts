import { Reducer } from 'redux';

import { SearchTypeChangedAction } from '../../domains/search/actions';
import {
  FetchSearchOptionsFulfilledAction,
  FetchSearchOptionsPendingAction,
  FetchSearchOptionsRejectedAction,
} from '../../domains/searchOptions/actions';

import { SearchOptions } from './types';

export type FetchSearchOptionsAction =
  | FetchSearchOptionsPendingAction
  | FetchSearchOptionsFulfilledAction
  | FetchSearchOptionsRejectedAction
  | SearchTypeChangedAction;

const initialState: SearchOptions = {
  isFetching: false,
  hasFailed: false,
  options: {},
};

const reducer: Reducer<SearchOptions, FetchSearchOptionsAction> = (
  state = initialState,
  action
) => {
  switch (action.type) {
    case 'FETCH_SEARCH_OPTIONS.PENDING':
      return { ...state, isFetching: true, hasFailed: false };
    case 'FETCH_SEARCH_OPTIONS.FULFILLED':
      return { ...state, isFetching: false, hasFailed: false, options: action.payload.options };
    case 'FETCH_SEARCH_OPTIONS.REJECTED':
      return { ...state, isFetching: false, hasFailed: true };
    case 'SEARCH.TYPE_CHANGED':
      return initialState;
    default: {
      return state;
    }
  }
};

export default reducer;
