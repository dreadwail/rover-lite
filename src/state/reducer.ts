import { combineReducers } from 'redux';

import results from './results/reducer';
import search from './search/reducer';
import searchOptions from './searchOptions/reducer';
import searchTypes from './searchTypes/reducer';
import { RoverSearchState } from './types';

export default () =>
  combineReducers<RoverSearchState>({
    searchTypes,
    searchOptions,
    search,
    results,
  });
