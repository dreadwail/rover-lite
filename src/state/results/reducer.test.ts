import {
  FetchResultsFulfilledAction,
  FetchResultsPendingAction,
  FetchResultsRejectedAction,
} from '../../domains/results/actions';
import { SearchFieldChangedAction, SearchTypeChangedAction } from '../../domains/search/actions';

import reducer, { ResultsAction } from './reducer';
import { Listing, Results } from './types';

const unhandledAction = ({ type: 'UNHANDLED_TEST_ACTION' } as any) as ResultsAction;

describe('results reducer', () => {
  let oldState: Results;
  let newState: Results;

  beforeEach(() => {
    oldState = {
      isFetching: false,
      hasFailed: false,
      isStale: false,
      listings: [],
    };
  });

  describe('when passed no previous state', () => {
    beforeEach(() => {
      oldState = undefined as any;
      newState = reducer(undefined, unhandledAction);
    });

    it('returns the correct initial state', () => {
      expect(newState).toMatchSnapshot();
    });
  });

  describe('when receiving an unhandled action', () => {
    beforeEach(() => {
      newState = reducer(oldState, unhandledAction);
    });

    it('returns the old state', () => {
      expect(newState).toEqual(oldState);
    });
  });

  describe('when fetching results starts', () => {
    beforeEach(() => {
      const action: FetchResultsPendingAction = { type: 'FETCH_RESULTS.PENDING' };
      newState = reducer(oldState, action);
    });

    it('returns the correct state', () => {
      expect(newState).toMatchSnapshot();
    });
  });

  describe('when fetching results succeeds', () => {
    beforeEach(() => {
      const testListing: Listing = {
        thumbnailUrl: 'https://example.com/profile.png',
        rank: 0,
        name: 'Test Name',
        badges: [
          {
            content: 'This sitter is Certified as Awesome.',
            imageUrl: 'https://example.com/badge.png',
            title: 'Certified Awesome',
          },
        ],
        tagline: 'Enthusiastic dog lover in Ballard.',
        neighborhood: 'Ballard',
        city: 'Seattle',
        state: 'WA',
        zip: '98117',
        repeatClientCount: 42,
        reviewCount: 12,
        ratingAverage: 4.8,
        featuredReview: 'This sitter is awesome.',
        currencySymbol: '$',
        price: 42.12,
        priceUnit: 'night',
        webUrl: 'https://www.rover.com/members/not-a-real-profile-just-a-test/',
      };
      const listings: Listing[] = [testListing];
      const action: FetchResultsFulfilledAction = {
        type: 'FETCH_RESULTS.FULFILLED',
        payload: listings,
      };
      newState = reducer(oldState, action);
    });

    it('returns the correct state', () => {
      expect(newState).toMatchSnapshot();
    });
  });

  describe('when fetching results fails', () => {
    const error = new Error('bang');

    beforeEach(() => {
      const action: FetchResultsRejectedAction = {
        type: 'FETCH_RESULTS.REJECTED',
        payload: {
          error,
        },
        error: true,
      };
      newState = reducer(oldState, action);
    });

    it('returns the correct state', () => {
      expect(newState).toMatchSnapshot();
    });
  });

  describe('when the search type changes', () => {
    beforeEach(() => {
      const action: SearchTypeChangedAction = {
        type: 'SEARCH.TYPE_CHANGED',
        payload: 'overnight-boarding',
      };
      newState = reducer(oldState, action);
    });

    it('returns the correct state', () => {
      expect(newState).toMatchSnapshot();
    });
  });

  describe('when a search field changes', () => {
    beforeEach(() => {
      const action: SearchFieldChangedAction = {
        type: 'SEARCH.FIELD_CHANGED',
        payload: {
          key: 'test-key',
          value: 'test-value',
        },
      };
      newState = reducer(oldState, action);
    });

    it('returns the correct state', () => {
      expect(newState).toMatchSnapshot();
    });
  });
});
