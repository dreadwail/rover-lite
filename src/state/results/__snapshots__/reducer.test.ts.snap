// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`results reducer when a search field changes returns the correct state 1`] = `
Object {
  "hasFailed": false,
  "isFetching": false,
  "isStale": true,
  "listings": Array [],
}
`;

exports[`results reducer when fetching results fails returns the correct state 1`] = `
Object {
  "hasFailed": true,
  "isFetching": false,
  "isStale": false,
  "listings": Array [],
}
`;

exports[`results reducer when fetching results starts returns the correct state 1`] = `
Object {
  "hasFailed": false,
  "isFetching": true,
  "isStale": false,
  "listings": Array [],
}
`;

exports[`results reducer when fetching results succeeds returns the correct state 1`] = `
Object {
  "hasFailed": false,
  "isFetching": false,
  "isStale": false,
  "listings": Array [
    Object {
      "badges": Array [
        Object {
          "content": "This sitter is Certified as Awesome.",
          "imageUrl": "https://example.com/badge.png",
          "title": "Certified Awesome",
        },
      ],
      "city": "Seattle",
      "currencySymbol": "$",
      "featuredReview": "This sitter is awesome.",
      "name": "Test Name",
      "neighborhood": "Ballard",
      "price": 42.12,
      "priceUnit": "night",
      "rank": 0,
      "ratingAverage": 4.8,
      "repeatClientCount": 42,
      "reviewCount": 12,
      "state": "WA",
      "tagline": "Enthusiastic dog lover in Ballard.",
      "thumbnailUrl": "https://example.com/profile.png",
      "webUrl": "https://www.rover.com/members/not-a-real-profile-just-a-test/",
      "zip": "98117",
    },
  ],
}
`;

exports[`results reducer when passed no previous state returns the correct initial state 1`] = `
Object {
  "hasFailed": false,
  "isFetching": false,
  "isStale": true,
  "listings": Array [],
}
`;

exports[`results reducer when the search type changes returns the correct state 1`] = `
Object {
  "hasFailed": false,
  "isFetching": false,
  "isStale": true,
  "listings": Array [],
}
`;
